src/Compilers/CSharp/Portable/Symbols/NamedTypeSymbol.cs

```diff
@@ -599,7 +599,10 @@ bool tryCalculateRequiredMembers(out ImmutableSegmentedDictionary<string, Symbol
                 var baseAllRequiredMembers = BaseTypeNoUseSiteDiagnostics?.AllRequiredMembers ?? ImmutableSegmentedDictionary<string, Symbol>.Empty;
                 var hasDeclaredRequiredMembers = HasDeclaredRequiredMembers;
 
-                foreach (var member in GetMembersUnordered())
+                // Iterate over members of the original definition to avoid realizing GetMembersUnordered() for
+                // substituted symbols. Substituted symbols are only realized for uncommon cases where they are
+                // necessary for checks against members from the base type.
+                foreach (var member in OriginalDefinition.GetMembersUnordered())
```

Comment:
>OriginalDefinition.GetMembersUnordered() [](http://example.com/codeflow?start=39&length=40)

It looks like for long tuples looking at the original definition members is going to miss fields past `Item7`, and behavior for scenario like the following will change (the error for `Item9` will disappear):
```
    [Fact]
    public void TupleWithRequiredFields_TupleExpressonSyntax_03()
    {
        var comp = CreateCompilation(new[] { """
            #pragma warning disable CS0219 // Unused local
            (int, int, int, int, int, int, int, double, float) t3 = new() {Item8 = 1};
            """, TupleWithRequiredMemberDefinition(false) }, targetFramework: TargetFramework.Mscorlib461 /* Using 461 to get a framework without ValueTuple */);

        comp.VerifyEmitDiagnostics(
            // 0.cs(2,57): error CS9035: Required member '(int, int, int, int, int, int, int, double, float).Item9' must be set in the object initializer or attribute constructor.
            // (int, int, int, int, int, int, int, double, float) t3 = new() {Item8 = 1};
            Diagnostic(ErrorCode.ERR_RequiredMemberMustBeSet, "new").WithArguments("(int, int, int, int, int, int, int, double, float).Item9").WithLocation(2, 57),
            // 0.cs(2,57): error CS9035: Required member '(int, int, int, int, int, int, int, double, float).Rest' must be set in the object initializer or attribute constructor.
            // (int, int, int, int, int, int, int, double, float) t3 = new() {Item8 = 1};
            Diagnostic(ErrorCode.ERR_RequiredMemberMustBeSet, "new").WithArguments("(int, int, int, int, int, int, int, double, float).Rest").WithLocation(2, 57)
            );
    }
```

However, it is not clear what the expected behavior should be in this case:
- There are no errors for `AnotherField` and `Property`. Is this expected?
- Perhaps the error for `Item9` is unexpected instead?
- Should a target typed `new` like that be even allowed? We don't allow `new` expressions with explicit tuple types (error CS8181: 'new' cannot be used with tuple type. Use a tuple literal expression instead.)

Depending on what we think the expected behavior should be, stopping treating items past `Item7` as required might not be a bad thing (unobservable), or even might be a good thing (observable).

Reply:
1. I don't think this is expected. It feels like there should be errors for these two unrelated properties.
2. I do think that's expected.
3. Probably not, though that does feel like an unrelated oversight. `(int, int) t = new() { Item1 = 1, Item2 = 2 };` compiles today. @jcouv, what do you think about that? Clearly, it's treating the `new()` as if it was `new ValueTuple<int, int>`, rather than `new (int, int)()`, but I'm unsure if that was intentional.

Reply:
What behavior do we expect for
```
    [Fact]
    public void TupleWithRequiredFields_TupleExpressonSyntax_04()
    {
        var comp = CreateCompilation(new[] { """
            #pragma warning disable CS0219 // Unused local
            var t3 = new System.ValueTuple<int, int, int, int, int, int, int, System.ValueTuple<double, float>>() {Item8 = 1};
            """, TupleWithRequiredMemberDefinition(false) }, targetFramework: TargetFramework.Mscorlib461 /* Using 461 to get a framework without ValueTuple */);

        comp.VerifyEmitDiagnostics(
            // 0.cs(2,14): error CS9035: Required member '(int, int, int, int, int, int, int, double, float).Item9' must be set in the object initializer or attribute constructor.
            // var t3 = new System.ValueTuple<int, int, int, int, int, int, int, System.ValueTuple<double, float>>() {Item8 = 1};
            Diagnostic(ErrorCode.ERR_RequiredMemberMustBeSet, "System.ValueTuple<int, int, int, int, int, int, int, System.ValueTuple<double, float>>").WithArguments("(int, int, int, int, int, int, int, double, float).Item9").WithLocation(2, 14),
            // 0.cs(2,14): error CS9035: Required member '(int, int, int, int, int, int, int, double, float).Rest' must be set in the object initializer or attribute constructor.
            // var t3 = new System.ValueTuple<int, int, int, int, int, int, int, System.ValueTuple<double, float>>() {Item8 = 1};
            Diagnostic(ErrorCode.ERR_RequiredMemberMustBeSet, "System.ValueTuple<int, int, int, int, int, int, int, System.ValueTuple<double, float>>").WithArguments("(int, int, int, int, int, int, int, double, float).Rest").WithLocation(2, 14)
            );
    }
```

Reply:
For the standard `new ValueTuple` case, I think `Item9` is probably unexpected. And, given that point, I think it does seem like my previous answers are probably inconsistent.

After some more consideration, I think the explicit `new` (whether inferred or not) should be consistent with non-tuple type behavior. That would mean that `Item9` is unexpected here, and I would not expect to see errors `AnotherField` or `AnotherProperty`, as those aren't on `ValueTuple<T1..., TRest>`. Those errors would surface when doing `TRest = new() { }` in a nested object initializer.